relay-common/src/utils.rs

```diff
@@ -1,11 +1,49 @@
 use std::fmt;
+use std::mem;
 use std::str;
+use std::sync::{Arc, Mutex};
 
 use once_cell::sync::OnceCell;
 use regex::Regex;
 
 use crate::macros::impl_str_serde;
 
+/// Wraps either the provided pattern string, which will be consumed and replaced with compiled
+/// regex, when the `compiled` function is called.
+#[derive(Debug, Clone)]
+struct GlobRegex {
+    // Note: Mutex and Arc here used to make sure this struct can be Send and Sync. Also that we
+    // could use interior mutabilty in order to be able to consume the value when needed.
+    value: Arc<Mutex<String>>,
+    re: OnceCell<Regex>,
+}
+
+impl GlobRegex {
+    /// Creates a new `GlobRegex` from the string pattern.
+    fn new(pattern: String) -> Self {
+        Self {
+            value: Arc::new(Mutex::new(pattern)),
+            re: OnceCell::new(),
+        }
+    }
+
+    /// Returns the either `Some` of the compiled regex, or `None` if there is an error while the
+    /// regex was created.
+    ///
+    /// This will compile regex once and will reuse it later on consecutive calls.
+    ///
+    /// By compiling the regex we consume value pattern, replacing it with empty string buffer,
+    /// which is very inexpensive operation.
+    fn compiled(&self) -> Option<&Regex> {
+        self.re
+            .get_or_try_init(|| {
+                let pattern = mem::take(&mut *self.value.lock().unwrap());
+                Regex::new(&pattern)
+            })
```

Comment:
How does `Regex::new` behave if the pattern is invalid? Does it panic (and the panic is propagated by `get_or_try_init`)? Do we get any information (e.g. Sentry issue) if relay fails to build the regex from a project config?

Reply:
[`Regex::new`](https://docs.rs/regex/latest/regex/struct.Regex.html#method.new) returns a `Result`.